File: app/src/main/java/com/meenbeese/chronos/fragments/sound/FileSoundChooserFragment.java
package com.meenbeese.chronos.fragments.sound;

import static android.app.Activity.RESULT_OK;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import com.meenbeese.chronos.R;
import com.meenbeese.chronos.activities.FileChooserActivity;
import com.meenbeese.chronos.adapters.SoundsAdapter;
import com.meenbeese.chronos.data.SoundData;
import com.meenbeese.chronos.fragments.BasePagerFragment;
import com.meenbeese.chronos.interfaces.SoundChooserListener;


public class FileSoundChooserFragment extends BaseSoundChooserFragment {

    private static final int REQUEST_AUDIO = 285;

    private static final String TYPE_AUDIO = "audio/*";

    private static final String SEPARATOR = ":ChronosFileSound:";
    private static final String PREF_FILES = "previousFiles";

    private SharedPreferences prefs;
    private List<SoundData> sounds;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_sound_chooser_file, container, false);

        prefs = PreferenceManager.getDefaultSharedPreferences(getContext());

        view.findViewById(R.id.addAudioFile).setOnClickListener(v -> launchFileChooser());

        RecyclerView recycler = view.findViewById(R.id.recycler);

        List<String> previousFiles = new ArrayList<>(prefs.getStringSet(PREF_FILES, new HashSet<>()));
        previousFiles.sort((o1, o2) -> {
            try {
                return Integer.parseInt(o1.split(SEPARATOR)[0]) - Integer.parseInt(o2.split(SEPARATOR)[0]);
            } catch (NumberFormatException e) {
                return 0;
            }
        });

        sounds = new ArrayList<>();
        for (String string : previousFiles) {
            String[] parts = string.split(SEPARATOR);
            sounds.add(new SoundData(parts[1], SoundData.TYPE_RINGTONE, parts[2]));
        }

        recycler.setLayoutManager(new LinearLayoutManager(getContext()));

        SoundsAdapter adapter = new SoundsAdapter(getChronos(), sounds);
        adapter.setListener(this);
        recycler.setAdapter(adapter);

        return view;
    }

    private void launchFileChooser() {
        Intent intent = new Intent(getContext(), FileChooserActivity.class);
        intent.putExtra(FileChooserActivity.EXTRA_TYPE, FileSoundChooserFragment.TYPE_AUDIO);
        startActivityForResult(intent, REQUEST_AUDIO);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == REQUEST_AUDIO && resultCode == RESULT_OK && data != null) {
            String name = "Audio File";
            if (data.hasExtra("name"))
                name = data.getStringExtra("name");

            onSoundChosen(new SoundData(name, SoundData.TYPE_RINGTONE, data.getDataString()));
        }
    }

    @Override
    public void onSoundChosen(SoundData sound) {
        super.onSoundChosen(sound);

        if (sound != null) {
            sounds.remove(sound);
            sounds.add(0, sound);

            Set<String> files = new HashSet<>();
            for (int i = 0; i < sounds.size(); i++) {
                files.add(i + SEPARATOR + sounds.get(i).getName() + SEPARATOR + sounds.get(i).getUrl());
            }

            prefs.edit().putStringSet(PREF_FILES, files).apply();
        }
    }



    @Override
    public String getTitle(Context context) {
        assert context != null;
        return context.getString(R.string.title_files);
    }

    public static class Instantiator extends BaseSoundChooserFragment.Instantiator {

        public Instantiator(Context context, SoundChooserListener listener) {
            super(context, listener);
        }

        @Override
        public BasePagerFragment newInstance(int position, SoundChooserListener listener) {
            BaseSoundChooserFragment fragment = new FileSoundChooserFragment();
            fragment.setListener(listener);
            return fragment;
        }

        @Override
        public String getTitle(Context context, int position) {
            assert context != null;
            return context.getString(R.string.title_files);
        }
    }
}

File: app/src/main/java/com/meenbeese/chronos/fragments/sound/FileSoundChooserFragment.kt
package com.meenbeese.chronos.fragments.sound

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.unit.dp
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringSetPreferencesKey
import androidx.fragment.app.FragmentActivity
import androidx.media3.common.util.UnstableApi

import com.meenbeese.chronos.R
import com.meenbeese.chronos.data.SoundData
import com.meenbeese.chronos.ext.dataStore
import com.meenbeese.chronos.fragments.BasePagerFragment
import com.meenbeese.chronos.fragments.FileChooserFragment
import com.meenbeese.chronos.interfaces.SoundChooserListener
import com.meenbeese.chronos.utils.AudioUtils
import com.meenbeese.chronos.views.SoundItemView

import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch

@UnstableApi
class FileSoundChooserFragment : BaseSoundChooserFragment() {
    private var sounds by mutableStateOf<List<SoundData>>(emptyList())
    private val PREF_FILES_KEY = stringSetPreferencesKey("previousFiles")
    private lateinit var audioUtils: AudioUtils

    @OptIn(DelicateCoroutinesApi::class)
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        audioUtils = AudioUtils(requireContext())

        return ComposeView(requireContext()).apply {
            setContent {
                MaterialTheme {
                    LaunchedEffect(Unit) {
                        val previousFiles = loadPreviousFiles()
                        val items = previousFiles.mapNotNull { str ->
                            val parts = str.split(SEPARATOR)
                            if (parts.size == 3) SoundData(parts[1], SoundData.TYPE_RINGTONE, parts[2]) else null
                        }
                        sounds = items
                    }

                    var currentlyPlayingUrl by remember { mutableStateOf<String?>(null) }

                    Column(modifier = Modifier.padding(16.dp)) {
                        Button(
                            onClick = { launchFileChooser() },
                            modifier = Modifier.fillMaxWidth()
                        ) {
                            Text("Add Audio File")
                        }

                        Spacer(modifier = Modifier.height(16.dp))

                        LazyColumn(modifier = Modifier.fillMaxSize()) {
                            items(sounds) { sound ->
                                val isPlaying = currentlyPlayingUrl == sound.url

                                SoundItemView(
                                    title = sound.name,
                                    isPlaying = isPlaying,
                                    modifier = Modifier.padding(vertical = 4.dp),
                                    onIconClick = {
                                        if (isPlaying) {
                                            audioUtils.stopCurrentSound()
                                            currentlyPlayingUrl = null
                                        } else {
                                            audioUtils.stopCurrentSound()
                                            audioUtils.playStream(sound.url, sound.type, null)
                                            currentlyPlayingUrl = sound.url
                                        }
                                    }
                                )
                            }
                        }
                    }
                }
            }
        }
    }

    private fun launchFileChooser() {
        val fragment = FileChooserFragment.newInstance(null, TYPE_AUDIO)
        fragment.setCallback { name, uri ->
            onSoundChosen(SoundData(name, SoundData.TYPE_RINGTONE, uri))
        }

        val activity = context as FragmentActivity
        activity.supportFragmentManager.beginTransaction()
            .replace(android.R.id.content, fragment)
            .addToBackStack(null)
            .commit()
    }

    @OptIn(DelicateCoroutinesApi::class)
    override fun onSoundChosen(sound: SoundData?) {
        super.onSoundChosen(sound)
        sound?.let {
            sounds = (sounds - it).toMutableList().apply { add(0, it) }
            GlobalScope.launch {
                savePreviousFiles()
            }
        }
    }

    private suspend fun loadPreviousFiles(): List<String> {
        val preferences = context?.dataStore?.data?.first()
        val previousFiles = preferences?.get(PREF_FILES_KEY) ?: emptySet()
        return previousFiles.toList().sortedWith { o1, o2 ->
            try {
                Integer.parseInt(o1.split(SEPARATOR.toRegex())[0]) - Integer.parseInt(o2.split(SEPARATOR.toRegex())[0])
            } catch (_: NumberFormatException) {
                0
            }
        }
    }

    private suspend fun savePreviousFiles() {
        val files: MutableSet<String> = HashSet()
        for (i in sounds.indices) {
            files.add(i.toString() + SEPARATOR + sounds[i].name + SEPARATOR + sounds[i].url)
        }

        context?.dataStore?.edit { preferences ->
            preferences[PREF_FILES_KEY] = files
        }
    }

    override fun getTitle(context: Context?): String? {
        return context?.getString(R.string.title_files)
    }

    internal class Instantiator(context: Context?, listener: SoundChooserListener?) :
        BaseSoundChooserFragment.Instantiator(context, listener) {
        override fun newInstance(
            position: Int,
            listener: SoundChooserListener?
        ): BasePagerFragment {
            val fragment: BaseSoundChooserFragment = FileSoundChooserFragment()
            fragment.setListener(listener)
            return fragment
        }

        override fun getTitle(context: Context?, position: Int): String? {
            return context?.getString(R.string.title_files)
        }
    }

    companion object {
        private const val TYPE_AUDIO = "audio/*"
        private const val SEPARATOR = ":ChronosFileSound:"
    }
}
